### ISSUE TITLE
UI5 tooling fails in nested workspaces

### ISSUE THREAD

**Issue by Akatuoro** [SAP/ui5-tooling#1033]
## Expected Behavior

Tooling works in typical npm dev environments, including nested workspaces.

## Current Behavior

Using the `cds-plugin-ui5` package, an error occurs in the `@ui5/project` loading node dependencies:

```
node_modules/@ui5/project/lib/graph/providers/NodePackageDependencies.js:105
                        throw new Error(
                              ^

Error: Unable to locate module @sap-ux/eslint-plugin-fiori-tools via resolve logic: Cannot find module '@sap-ux/eslint-plugin-fiori-tools/package.json' from 'cap/sflight/app/travel_analytics'
    at NodePackageDependencies._resolveModulePath (node_modules/@ui5/project/lib/graph/providers/NodePackageDependencies.js:105:10)
    at async node_modules/@ui5/project/lib/graph/providers/NodePackageDependencies.js:74:23
    at async Promise.all (index 0)
    at async projectGraphBuilder (node_modules/@ui5/project/lib/graph/projectGraphBuilder.js:150:27)
    at async graphFromPackageDependencies (node_modules/@ui5/project/lib/graph/graph.js:73:23)
    at async applyUI5Middleware (node_modules/cds-plugin-ui5/lib/applyUI5Middleware.js:69:16)
    at async cds.bootstrap (node_modules/cds-plugin-ui5/cds-plugin.js:188:22)
```

This seems to be using some custom module resolution... tapping into the node module resolution via `require.resolve` or `import.meta.resolve` would likely be more robust...

## Steps to Reproduce the Issue

1. git clone https://github.com/Akatuoro/ui5-tooling-dev-test
2. npm i
3. cd cap-sflight
4. npx cds-typer "*"
5. cds watch

> if cds is not installed, install it via npm i -g @sap/cds-dk

## Context
 * UI5 Module Version (output of `ui5 --version` when using the CLI): **`unknown`**
 * Node.js Version: **`v22.7.0`**
 * npm Version: `10.8.3`
 * OS/Platform: `macOS`

**Comment by matz3** [comment]
Thank you for reporting this and providing steps to reproduce.

I followed the steps but was unable to get to the reported error (using node v22.7.0 / npm v10.8.3):
```
~/github/Akatuoro/ui5-tooling-dev-test main
❯ npm i              
npm warn EBADENGINE Unsupported engine {
npm warn EBADENGINE   package: '@sap-ux/vocabularies-types@0.10.0',
npm warn EBADENGINE   required: { node: '>=18.0.0 < 19.0.0 || >=20.0.0 < 21.0.0', pnpm: '>=8' },
npm warn EBADENGINE   current: { node: 'v22.7.0', npm: '10.8.3' }
npm warn EBADENGINE }
npm warn EBADENGINE Unsupported engine {
npm warn EBADENGINE   package: '@sap-ux/vocabularies-types@0.10.0',
npm warn EBADENGINE   required: { node: '>=18.0.0 < 19.0.0 || >=20.0.0 < 21.0.0', pnpm: '>=8' },
npm warn EBADENGINE   current: { node: 'v22.7.0', npm: '10.8.3' }
npm warn EBADENGINE }
npm warn EBADENGINE Unsupported engine {
npm warn EBADENGINE   package: '@sap-ux/vocabularies-types@0.10.0',
npm warn EBADENGINE   required: { node: '>=18.0.0 < 19.0.0 || >=20.0.0 < 21.0.0', pnpm: '>=8' },
npm warn EBADENGINE   current: { node: 'v22.7.0', npm: '10.8.3' }
npm warn EBADENGINE }
npm warn deprecated inflight@1.0.6: This module is not supported, and leaks memory. Do not use it. Check out lru-cache if you want a good and tested way to coalesce async requests by a key value, which is much more comprehensive and powerful.
npm warn deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
npm warn deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
npm warn deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported
npm warn deprecated glob@7.2.3: Glob versions prior to v9 are no longer supported

added 941 packages, and audited 943 packages in 5s

140 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
                                                                                                                                   
~/github/Akatuoro/ui5-tooling-dev-test main
❯ cd cap-sflight
                                                                                                                                   
~/github/Akatuoro/ui5-tooling-dev-test/cap-sflight heads/main
❯ npx cds-typer "*"
[cds-plugin-ui5] [WARN] The cds build task requires @sap/cds-dk version >= 7.5.0! Skipping execution as your @sap/cds-dk version is too old...
```

After running `npm i -g @sap/cds-dk` I get the same error. `npm ls -g @sap/cds-dk` gives me `@sap/cds-dk@8.5.1` as output.

**Comment by Akatuoro** [comment]
@matz3 , thanks for trying it. That seems to be another issue in the `cds-plugin-ui5` :D 
But the warning message should have no effect here.

What is the output when you run `cds watch`? Or `cds-tsx watch` (<- needs `npm i -g tsx`)?

**Comment by matz3** [comment]
Ah okay. I just didn't continue as I expected that there was more output from `npx cds-typer "*"`.
I could reproduce it now and will have a look into it.

**Comment by matz3** [comment]
@Akatuoro the problem here is that there are two npm workspaces. One in the root (that includes `cap-sflight`), and one in the `cap-sflight` folder (that includes `app/*`). Up to my knowledge npm only supports one workspace, so it essentially ignores the second level (`app/*`), which causes the dependencies of those packages to be left out when running `npm install`.

I could verify that by running `npm ls @sap-ux/eslint-plugin-fiori-tools` in `ui5-tooling-dev-test/cap-sflight/app/travel_analytics` which gave me no installed version. Only after removing the outer workspace configuration (root `package.json`) and running `npm install` again (in `cap-sflight`) I was getting a version when checking the installation with `npm ls @sap-ux/eslint-plugin-fiori-tools` and also could successfully run `cds watch` as described in the steps.

We are using the [resolve](https://www.npmjs.com/package/resolve) package for the lookup, which is widely adopted and expected to give the same behavior as native resolution logic.

Our API/CLI fails in case a dependency from the UI5 project's `package.json` can not be resolved. This usually indicates that the project's dependencies haven't been installed, which is a requirement to run UI5 Tooling as usually middlewares or tasks are provided via dependencies (in this case `@sap/ux-ui5-tooling`). `@sap-ux/eslint-plugin-fiori-tools` just happens to be the first one that is checked, so it fails there.

**Comment by Akatuoro** [comment]
@matz3 , you're right, this seems to be an issue in npm since 8.5.0: https://github.com/npm/cli/issues/4774

A workaround is to go into `app/travel_processor` and call `npm i` there... then the corresponding node modules are actually installed in the `cap-sflight` folder and `cds watch` works with the ui5 tooling...

Closing this as it can only be fixed in npm. Thanks for analyzing it.
